sormas-ui/src/main/java/de/symeda/sormas/ui/labmessage/LabMessageController.java

```diff
@@ -622,13 +627,118 @@ private void editSample(SampleDto sample, LabMessageDto labMessage) {
+
 		sampleEditComponent.addCommitListener(window::close);
 		sampleEditComponent.addDiscardListener(window::close);
 
 		showFormWithLabMessage(labMessage, sampleEditComponent, window, I18nProperties.getString(Strings.headingEditSample), false);
 
 	}
 
+	private void addReferOrLinkToOtherLabButton(
+		CommitDiscardWrapperComponent<SampleEditForm> editForm,
+		Disease disease,
+		LabMessageDto labMessage,
+		Window window) {
+		Button referOrLinkToOtherLabButton = null;
+		SampleDto sample = editForm.getWrappedComponent().getValue();
+		if (sample.getReferredTo() == null) {
+			if (UserProvider.getCurrent().hasUserRight(UserRight.SAMPLE_TRANSFER)) {
+				referOrLinkToOtherLabButton =
+					ButtonHelper.createButton("referOrLinkToOtherLab", I18nProperties.getCaption(Captions.sampleRefer), new Button.ClickListener() {
+
+						private static final long serialVersionUID = 1L;
+
+						@Override
+						public void buttonClick(Button.ClickEvent event) {
+							try {
+								editForm.getWrappedComponent().commit();
+								SampleDto sampleDto = editForm.getWrappedComponent().getValue();
+								sampleDto = FacadeProvider.getSampleFacade().saveSample(sampleDto);
+								createSampleReferral(sampleDto, disease, labMessage);
+								window.close();
+							} catch (Buffered.SourceException | Validator.InvalidValueException e) {
+								Notification.show(I18nProperties.getString(Strings.messageSampleErrors), Notification.Type.ERROR_MESSAGE);
+							}
+						}
+					}, ValoTheme.BUTTON_LINK);
+			}
+		} else {
+			SampleDto referredDto = FacadeProvider.getSampleFacade().getSampleByUuid(sample.getReferredTo().getUuid());
+			FacilityReferenceDto referredDtoLab = referredDto.getLab();
+			String referOrLinkToOtherLabButtonCaption = referredDtoLab == null
+				? I18nProperties.getCaption(Captions.sampleReferredToInternal) + " ("
+					+ DateFormatHelper.formatLocalDateTime(referredDto.getSampleDateTime()) + ")"
+				: I18nProperties.getCaption(Captions.sampleReferredTo) + " " + referredDtoLab.toString();
+
+			referOrLinkToOtherLabButton =
+				ButtonHelper.createButton("referOrLinkToOtherLab", referOrLinkToOtherLabButtonCaption, new Button.ClickListener() {
+
+					private static final long serialVersionUID = 1L;
+
+					@Override
+					public void buttonClick(Button.ClickEvent event) {
+						editSample(referredDto, labMessage);
+						window.close();
+					}
+				});
+		}
+
+		if (referOrLinkToOtherLabButton != null) {
+			editForm.getButtonsPanel().addComponentAsFirst(referOrLinkToOtherLabButton);
+			editForm.getButtonsPanel().setComponentAlignment(referOrLinkToOtherLabButton, Alignment.BOTTOM_LEFT);
+		}
+	}
+
+	public void addReferredFromButton(CommitDiscardWrapperComponent<SampleEditForm> editForm, LabMessageDto labMessage, Window window) {
+		SampleReferenceDto referredFromRef = FacadeProvider.getSampleFacade().getReferredFrom(editForm.getWrappedComponent().getValue().getUuid());
```

Comment:
seems to be a duplicate of SampleController:: addReferredFromButton

Reply:
Same as above.

Reply:
Adding Runnable or Consumer argument(s) on SampleController::addReferOrLinkToOtherLabButton to have different handlers would not increase the complexity but instead there won't be duplicated code for the same logic.

Reply:
The problem I see with this is that both Runnables would need to fetch the `referredFrom` SampleDto, so most of the code lines would still be duplicated.
Consumers would need a different amount of parameters for sample editing and lab message processing. Maybe we should wait for a third opinion from @vidi42 ?

Reply:
Sorry, I was wrong here. Consumers should work in this case. I will change to @leventegal-she suggestion.